File: packages/notes/src/__tests__/supabase-notes-adapter.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { NoteId, NotebookId, TagId, UserId } from "@notes-app/domain";
import { SupabaseNotesAdapter } from "../supabase-notes-adapter.js";
import { NotesNotFoundError, type NotesHttp } from "../notes-http.js";
import type { NoteRow, NotebookRow, TagRow, NoteHitRow } from "../rows.js";

const OWNER = UserId("550e8400-e29b-41d4-a716-446655440000");
const NOTE = NoteId("550e8400-e29b-41d4-a716-446655440001");
const NOTEBOOK = NotebookId("550e8400-e29b-41d4-a716-446655440002");
const TAG = TagId("550e8400-e29b-41d4-a716-446655440003");
const TAG_2 = TagId("550e8400-e29b-41d4-a716-446655440004");

function makeNoteRow(overrides: Partial<NoteRow> = {}): NoteRow {
  return {
    id: NOTE,
    owner_id: OWNER,
    notebook_id: NOTEBOOK,
    title: "Welcome",
    body_json: { type: "doc", content: [] },
    is_pinned: false,
    is_trashed: false,
    trashed_at: null,
    ai_excluded: false,
    created_at: "2026-05-20T12:00:00.000Z",
    updated_at: "2026-05-20T12:00:00.000Z",
    ...overrides,
  };
}

function makeNotebookRow(overrides: Partial<NotebookRow> = {}): NotebookRow {
  return {
    id: NOTEBOOK,
    owner_id: OWNER,
    name: "Inbox",
    color: "#aabbcc",
    sort_order: 0,
    created_at: "2026-05-20T12:00:00.000Z",
    updated_at: "2026-05-20T12:00:00.000Z",
    ...overrides,
  };
}

function makeTagRow(overrides: Partial<TagRow> = {}): TagRow {
  return {
    id: TAG,
    owner_id: OWNER,
    name: "work",
    color: "#112233",
    created_at: "2026-05-20T12:00:00.000Z",
    updated_at: "2026-05-20T12:00:00.000Z",
    ...overrides,
  };
}

function makeHttp(): NotesHttp {
  return {
    selectNotes: vi.fn(),
    selectNote: vi.fn(),
    insertNote: vi.fn(),
    patchNote: vi.fn(),
    setNoteTrashed: vi.fn(),
    selectNotebooks: vi.fn(),
    insertNotebook: vi.fn(),
    patchNotebook: vi.fn(),
    deleteNotebook: vi.fn(),
    selectTags: vi.fn(),
    insertTag: vi.fn(),
    mergeTag: vi.fn(),
    searchNotes: vi.fn(),
  } as unknown as NotesHttp;
}

describe("SupabaseNotesAdapter — notes", () => {
  let http: NotesHttp;
  let adapter: SupabaseNotesAdapter;
  beforeEach(() => {
    http = makeHttp();
    adapter = new SupabaseNotesAdapter(http);
  });

  it("createNote rejects a blank title before hitting the wire", async () => {
    await expect(
      adapter.createNote({
        ownerId: OWNER,
        title: "  ",
        bodyJson: "{}",
      }),
    ).rejects.toThrow(/title/);
    expect(http.insertNote).not.toHaveBeenCalled();
  });

  it("createNote rejects an unparseable bodyJson before hitting the wire", async () => {
    await expect(
      adapter.createNote({
        ownerId: OWNER,
        title: "hi",
        bodyJson: "not json",
      }),
    ).rejects.toThrow(/valid JSON/);
  });

  it("createNote translates the inserted row into a domain Note", async () => {
    vi.mocked(http.insertNote).mockResolvedValue(makeNoteRow({ title: "First" }));
    const note = await adapter.createNote({
      ownerId: OWNER,
      title: "First",
      bodyJson: "{}",
    });
    expect(note.title).toBe("First");
  });

  it("updateNote requires at least one field", async () => {
    await expect(adapter.updateNote(NOTE, {})).rejects.toThrow(/at least one field/);
  });

  it("updateNote rejects a blank title patch", async () => {
    await expect(adapter.updateNote(NOTE, { title: "  " })).rejects.toThrow(/title/);
  });

  it("trashNote and restoreNote toggle the flag", async () => {
    await adapter.trashNote(NOTE);
    expect(http.setNoteTrashed).toHaveBeenCalledWith(NOTE, true);
    await adapter.restoreNote(NOTE);
    expect(http.setNoteTrashed).toHaveBeenCalledWith(NOTE, false);
  });

  it("listNotes paginates and translates rows", async () => {
    vi.mocked(http.selectNotes).mockResolvedValue({
      items: [makeNoteRow(), makeNoteRow({ id: "550e8400-e29b-41d4-a716-446655440099" })],
      total: 2,
    });
    const result = await adapter.listNotes({}, { offset: 0, limit: 50 });
    expect(result.total).toBe(2);
    expect(result.items).toHaveLength(2);
  });

  it("listNotes rejects an invalid page", async () => {
    await expect(adapter.listNotes({}, { offset: -1, limit: 50 })).rejects.toThrow(/offset/);
    await expect(adapter.listNotes({}, { offset: 0, limit: 0 })).rejects.toThrow(/limit/);
    await expect(adapter.listNotes({}, { offset: 0, limit: 201 })).rejects.toThrow(/limit/);
  });

  it("getNote throws NotesNotFoundError on a 404", async () => {
    vi.mocked(http.selectNote).mockResolvedValue(null);
    await expect(adapter.getNote(NOTE)).rejects.toBeInstanceOf(NotesNotFoundError);
  });

  it("searchKeyword returns an empty array for a blank query without hitting the wire", async () => {
    const hits = await adapter.searchKeyword("   ", {});
    expect(hits).toEqual([]);
    expect(http.searchNotes).not.toHaveBeenCalled();
  });

  it("searchKeyword orders hits by rank descending", async () => {
    const rows: NoteHitRow[] = [
      { ...makeNoteRow({ id: "550e8400-e29b-41d4-a716-446655440010" }), snippet: "lo", rank: 0.1 },
      { ...makeNoteRow({ id: "550e8400-e29b-41d4-a716-446655440020" }), snippet: "hi", rank: 0.9 },
    ];
    vi.mocked(http.searchNotes).mockResolvedValue(rows);
    const hits = await adapter.searchKeyword("hello", {});
    expect(hits.map((h) => h.rank)).toEqual([0.9, 0.1]);
  });
});

describe("SupabaseNotesAdapter — notebooks", () => {
  let http: NotesHttp;
  let adapter: SupabaseNotesAdapter;
  beforeEach(() => {
    http = makeHttp();
    adapter = new SupabaseNotesAdapter(http);
  });

  it("createNotebook rejects a blank name", async () => {
    await expect(adapter.createNotebook({ ownerId: OWNER, name: "" })).rejects.toThrow(/name/);
  });

  it("renameNotebook rejects a blank name and otherwise patches", async () => {
    await expect(adapter.renameNotebook(NOTEBOOK, "")).rejects.toThrow(/name/);
    vi.mocked(http.patchNotebook).mockResolvedValue(makeNotebookRow({ name: "Renamed" }));
    const nb = await adapter.renameNotebook(NOTEBOOK, "Renamed");
    expect(nb.name).toBe("Renamed");
  });

  it("deleteNotebook forwards the move target", async () => {
    await adapter.deleteNotebook(NOTEBOOK, null);
    expect(http.deleteNotebook).toHaveBeenCalledWith(NOTEBOOK, null);
  });

  it("listNotebooks maps every row", async () => {
    vi.mocked(http.selectNotebooks).mockResolvedValue([
      makeNotebookRow(),
      makeNotebookRow({ id: "550e8400-e29b-41d4-a716-446655440099", name: "Archive" }),
    ]);
    const all = await adapter.listNotebooks();
    expect(all.map((n) => n.name)).toEqual(["Inbox", "Archive"]);
  });
});

describe("SupabaseNotesAdapter — tags", () => {
  let http: NotesHttp;
  let adapter: SupabaseNotesAdapter;
  beforeEach(() => {
    http = makeHttp();
    adapter = new SupabaseNotesAdapter(http);
  });

  it("createTag rejects a blank name", async () => {
    await expect(adapter.createTag({ ownerId: OWNER, name: "" })).rejects.toThrow(/name/);
  });

  it("mergeTag rejects merging a tag into itself", async () => {
    await expect(adapter.mergeTag(TAG, TAG)).rejects.toThrow(/itself/);
  });

  it("mergeTag delegates to the wire on distinct tags", async () => {
    await adapter.mergeTag(TAG, TAG_2);
    expect(http.mergeTag).toHaveBeenCalledWith(TAG, TAG_2);
  });

  it("listTags maps every row", async () => {
    vi.mocked(http.selectTags).mockResolvedValue([
      makeTagRow(),
      makeTagRow({ id: "550e8400-e29b-41d4-a716-446655440099", name: "home" }),
    ]);
    const all = await adapter.listTags();
    expect(all.map((t) => t.name)).toEqual(["work", "home"]);
  });
});

File: packages/android/notes-android/src/test/kotlin/app/notes/notes/SupabaseNotesAdapterTest.kt
package app.notes.notes

import app.notes.domain.Note
import app.notes.domain.NoteId
import app.notes.domain.Notebook
import app.notes.domain.NotebookId
import app.notes.domain.Tag
import app.notes.domain.TagId
import app.notes.domain.UserId
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.time.Instant

class SupabaseNotesAdapterTest {
    private val owner = UserId("550e8400-e29b-41d4-a716-446655440000")
    private val noteId = NoteId("550e8400-e29b-41d4-a716-446655440001")
    private val notebookId = NotebookId("550e8400-e29b-41d4-a716-446655440002")
    private val tag1 = TagId("550e8400-e29b-41d4-a716-446655440003")
    private val tag2 = TagId("550e8400-e29b-41d4-a716-446655440004")
    private val now = Instant.parse("2026-05-20T12:00:00Z")

    private fun note(title: String = "Welcome") = Note(
        id = noteId,
        ownerId = owner,
        notebookId = notebookId,
        title = title,
        bodyJson = "{\"type\":\"doc\",\"content\":[]}",
        tags = emptySet(),
        createdAt = now,
        updatedAt = now,
        isPinned = false,
        isTrashed = false,
        trashedAt = null,
        aiExcluded = false,
        attachmentRefs = emptyList(),
    )

    private fun notebook(name: String = "Inbox") = Notebook(
        id = notebookId,
        ownerId = owner,
        name = name,
        color = "#aabbcc",
        sortOrder = 0,
        createdAt = now,
        updatedAt = now,
    )

    private fun tag(id: TagId = tag1, name: String = "work") = Tag(
        id = id,
        ownerId = owner,
        name = name,
        color = "#112233",
        createdAt = now,
        updatedAt = now,
    )

    @Test
    fun `createNote rejects invalid bodyJson before hitting the wire`() = runTest {
        val http = FakeHttp(insertNoteResult = note())
        val adapter = SupabaseNotesAdapter(http)
        assertThrows(IllegalArgumentException::class.java) {
            // runBlocking-style via runTest captures the throw because validation is synchronous
            kotlinx.coroutines.runBlocking {
                adapter.createNote(NewNoteInput(owner, "hi", "not json"))
            }
        }
        assertEquals(0, http.insertNoteCalls)
    }

    @Test
    fun `createNote returns the note from the wire`() = runTest {
        val http = FakeHttp(insertNoteResult = note(title = "First"))
        val adapter = SupabaseNotesAdapter(http)
        val n = adapter.createNote(NewNoteInput(owner, "First", "{}"))
        assertEquals("First", n.title)
    }

    @Test
    fun `updateNote enforces NotePatch invariants`() {
        assertThrows(IllegalArgumentException::class.java) { NotePatch() }
        assertThrows(IllegalArgumentException::class.java) { NotePatch(title = "  ") }
    }

    @Test
    fun `updateNote rejects malformed bodyJson patch`() = runTest {
        val http = FakeHttp(patchNoteResult = note())
        val adapter = SupabaseNotesAdapter(http)
        assertThrows(IllegalArgumentException::class.java) {
            kotlinx.coroutines.runBlocking {
                adapter.updateNote(noteId, NotePatch(bodyJson = "lol"))
            }
        }
    }

    @Test
    fun `trash and restore toggle the wire`() = runTest {
        val http = FakeHttp()
        val adapter = SupabaseNotesAdapter(http)
        adapter.trashNote(noteId)
        adapter.restoreNote(noteId)
        assertEquals(listOf(true, false), http.trashCalls)
    }

    @Test
    fun `listNotes propagates the paged result`() = runTest {
        val http = FakeHttp(notes = listOf(note()), total = 1)
        val adapter = SupabaseNotesAdapter(http)
        val result = adapter.listNotes(NoteFilter(), Page(0, 50))
        assertEquals(1, result.total)
        assertEquals(1, result.items.size)
    }

    @Test
    fun `getNote throws NotesNotFoundException on 404`() {
        val http = FakeHttp(noteById = null)
        val adapter = SupabaseNotesAdapter(http)
        val ex = assertThrows(NotesNotFoundException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.getNote(noteId) }
        }
        assertSame(noteId, ex.id)
    }

    @Test
    fun `searchKeyword returns empty list for blank query without wire`() = runTest {
        val http = FakeHttp()
        val adapter = SupabaseNotesAdapter(http)
        val hits = adapter.searchKeyword("  ", NoteFilter())
        assertTrue(hits.isEmpty())
        assertEquals(0, http.searchCalls)
    }

    @Test
    fun `searchKeyword orders by rank descending`() = runTest {
        val http = FakeHttp(
            hits = listOf(
                NoteHit(note(), "low", 0.1),
                NoteHit(note(), "high", 0.9),
            ),
        )
        val adapter = SupabaseNotesAdapter(http)
        val hits = adapter.searchKeyword("hello", NoteFilter())
        assertEquals(listOf(0.9, 0.1), hits.map { it.rank })
    }

    @Test
    fun `renameNotebook rejects blank names`() {
        val http = FakeHttp(notebookPatched = notebook())
        val adapter = SupabaseNotesAdapter(http)
        assertThrows(IllegalArgumentException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.renameNotebook(notebookId, "  ") }
        }
    }

    @Test
    fun `mergeTag rejects merging into itself`() {
        val http = FakeHttp()
        val adapter = SupabaseNotesAdapter(http)
        assertThrows(IllegalArgumentException::class.java) {
            kotlinx.coroutines.runBlocking { adapter.mergeTag(tag1, tag1) }
        }
    }

    @Test
    fun `mergeTag delegates to the wire for distinct tags`() = runTest {
        val http = FakeHttp()
        val adapter = SupabaseNotesAdapter(http)
        adapter.mergeTag(tag1, tag2)
        assertEquals(listOf(tag1 to tag2), http.mergeCalls)
    }

    @Test
    fun `createTag and createNotebook enforce non-empty names at the input layer`() {
        assertThrows(IllegalArgumentException::class.java) {
            NewTagInput(owner, "  ")
        }
        assertThrows(IllegalArgumentException::class.java) {
            NewNotebookInput(owner, "  ")
        }
    }

    @Test
    fun `Page enforces bounds`() {
        assertThrows(IllegalArgumentException::class.java) { Page(-1, 50) }
        assertThrows(IllegalArgumentException::class.java) { Page(0, 0) }
        assertThrows(IllegalArgumentException::class.java) { Page(0, 201) }
    }

    private inner class FakeHttp(
        val insertNoteResult: Note = note(),
        val patchNoteResult: Note = note(),
        val notes: List<Note> = emptyList(),
        val total: Int = 0,
        val noteById: Note? = note(),
        val hits: List<NoteHit> = emptyList(),
        val notebookPatched: Notebook = notebook(),
    ) : NotesHttp {
        var insertNoteCalls = 0
        val trashCalls = mutableListOf<Boolean>()
        var searchCalls = 0
        val mergeCalls = mutableListOf<Pair<TagId, TagId>>()

        override suspend fun selectNotes(filter: NoteFilter, page: Page): PagedResult<Note> =
            PagedResult(notes, total)

        override suspend fun selectNote(id: NoteId): Note? = noteById

        override suspend fun insertNote(input: NewNoteInput): Note {
            insertNoteCalls += 1
            return insertNoteResult
        }

        override suspend fun patchNote(id: NoteId, patch: NotePatch): Note = patchNoteResult
        override suspend fun setNoteTrashed(id: NoteId, isTrashed: Boolean) { trashCalls.add(isTrashed) }

        override suspend fun selectNotebooks(): List<Notebook> = listOf(notebook())
        override suspend fun insertNotebook(input: NewNotebookInput): Notebook = notebook()
        override suspend fun patchNotebook(id: NotebookId, name: String): Notebook = notebookPatched
        override suspend fun deleteNotebook(id: NotebookId, moveNotesTo: NotebookId?) {}

        override suspend fun selectTags(): List<Tag> = listOf(tag())
        override suspend fun insertTag(input: NewTagInput): Tag = tag()
        override suspend fun mergeTag(from: TagId, into: TagId) { mergeCalls.add(from to into) }

        override suspend fun searchNotes(q: String, filter: NoteFilter): List<NoteHit> {
            searchCalls += 1
            return hits
        }
    }
}
